File: r/scripts/Run_Full_Analysis.R
# ==============================================================================
# Master Script: Run Complete Bayesian MCMC Analysis
# ==============================================================================
# This script orchestrates the complete analysis pipeline:
# 1. Data preprocessing
# 2. Baseline Gibbs sampling analysis
# 3. Metropolis-Hastings analysis
# 4. Algorithm comparison
# 5. Residual diagnostics
# 6. Comprehensive reporting
# ==============================================================================

# Set up paths
setwd(dirname(sys.frame(1)$ofile))  # Set working directory to script location

# Load required libraries
suppressPackageStartupMessages({
  library(MASS)
  library(coda)
  library(ggplot2)
  library(gridExtra)
})

# Source all analysis scripts
source("Data_Preprocessing.R")
source("Model_Setup.R")
source("Gibbs_Sampling.R")
source("Metropolis_Hastings.R")
source("Convergence_Detection.R")
source("Posterior_Inference.R")
source("Algorithm_Comparison.R")
source("Residual_Diagnostics.R")

cat("\n")
cat("╔"*80, "\n", sep="")
cat("║ BAYESIAN MCMC ANALYSIS: MEDICAL INSURANCE COSTS\n")
cat("║ Complete Analysis Pipeline\n")
cat("╚"*80, "\n\n", sep="")

# Start timing
analysis_start_time <- Sys.time()

# ==============================================================================
# STEP 1: DATA LOADING AND PREPARATION
# ==============================================================================

cat("━"*80, "\n", sep="")
cat("STEP 1: DATA LOADING AND PREPARATION\n")
cat("━"*80, "\n\n", sep="")

# Load cleaned data
data_path <- "../../data/expenses_cleaned.csv"
df <- read.csv(data_path)

cat(sprintf("✓ Loaded %d observations from %s\n", nrow(df), basename(data_path)))

# Prepare matrices
y <- df$charges
X <- cbind(1, as.matrix(df[, c("age", "sex", "bmi", "children", "smoker")]))
colnames(X) <- c("Intercept", "Age", "Sex", "BMI", "Children", "Smoker")

n <- nrow(X)
p <- ncol(X)

cat(sprintf("✓ Response variable: charges (n = %d)\n", n))
cat(sprintf("✓ Features: %d predictors + intercept\n", p-1))

# Descriptive statistics
cat(sprintf("\nData summary:\n"))
cat(sprintf("  Mean charges: $%.2f\n", mean(y)))
cat(sprintf("  SD charges: $%.2f\n", sd(y)))
cat(sprintf("  Range: $%.2f - $%.2f\n\n", min(y), max(y)))

# ==============================================================================
# STEP 2: GIBBS SAMPLING (BASELINE MODEL)
# ==============================================================================

cat("━"*80, "\n", sep="")
cat("STEP 2: GIBBS SAMPLING (BASELINE MODEL)\n")
cat("━"*80, "\n\n", sep="")

# Priors
prior_beta_mean <- rep(0, p)
prior_beta_precision <- diag(0.001, p)
prior_sigma2_shape <- 0.01
prior_sigma2_scale <- 0.01

# Sampling parameters
n_iter <- 50000
warmup <- 10000
n_chains <- 3

cat(sprintf("Running Gibbs sampler:\n"))
cat(sprintf("  - Iterations: %d (warmup: %d)\n", n_iter, warmup))
cat(sprintf("  - Chains: %d\n", n_chains))

set.seed(42)

gibbs_results <- lapply(1:n_chains, function(chain_id) {
  cat(sprintf("  Chain %d/%d...\n", chain_id, n_chains))
  gibbs_lm(y, X, prior_beta_mean, prior_beta_precision,
           prior_sigma2_shape, prior_sigma2_scale,
           n_iter, warmup, seed = 42 + chain_id)
})

cat("\n✓ Gibbs sampling complete\n\n")

# ==============================================================================
# STEP 3: METROPOLIS-HASTINGS ANALYSIS
# ==============================================================================

cat("━"*80, "\n", sep="")
cat("STEP 3: METROPOLIS-HASTINGS ANALYSIS\n")
cat("━"*80, "\n\n", sep="")

cat(sprintf("Running Metropolis-Hastings sampler:\n"))
cat(sprintf("  - Iterations: %d (warmup: %d)\n", n_iter, warmup))
cat(sprintf("  - Chains: %d\n", n_chains))
cat(sprintf("  - Adaptive tuning: enabled\n"))

set.seed(42)

mh_results <- lapply(1:n_chains, function(chain_id) {
  cat(sprintf("  Chain %d/%d...\n", chain_id, n_chains))
  metropolis_hastings_lm(y, X, prior_beta_mean, prior_beta_precision,
                        prior_sigma2_shape, prior_sigma2_scale,
                        n_iter, warmup, seed = 42 + chain_id)
})

cat("\n✓ Metropolis-Hastings sampling complete\n\n")

# ==============================================================================
# STEP 4: CONVERGENCE DIAGNOSTICS
# ==============================================================================

cat("━"*80, "\n", sep="")
cat("STEP 4: CONVERGENCE DIAGNOSTICS\n")
cat("━"*80, "\n\n", sep="")

# Check Gibbs convergence
cat("Gibbs Sampling Convergence:\n")
gibbs_beta_chains <- lapply(gibbs_results, function(x) x$beta)
gibbs_sigma2_chains <- lapply(gibbs_results, function(x) x$sigma2)

rhat_gibbs_beta <- sapply(1:p, function(j) {
  chains <- lapply(gibbs_beta_chains, function(x) x[, j])
  mcmc_list <- mcmc.list(lapply(chains, mcmc))
  gelman.diag(mcmc_list, autoburnin = FALSE)$psrf[1, 1]
})

for (j in 1:p) {
  status <- ifelse(rhat_gibbs_beta[j] < 1.1, "✓", "✗")
  cat(sprintf("  %s %s: R-hat = %.4f\n", status, colnames(X)[j], rhat_gibbs_beta[j]))
}

# Check MH convergence
cat("\nMetropolis-Hastings Convergence:\n")
mh_beta_chains <- lapply(mh_results, function(x) x$beta)
mh_sigma2_chains <- lapply(mh_results, function(x) x$sigma2)

rhat_mh_beta <- sapply(1:p, function(j) {
  chains <- lapply(mh_beta_chains, function(x) x[, j])
  mcmc_list <- mcmc.list(lapply(chains, mcmc))
  gelman.diag(mcmc_list, autoburnin = FALSE)$psrf[1, 1]
})

for (j in 1:p) {
  status <- ifelse(rhat_mh_beta[j] < 1.1, "✓", "✗")
  cat(sprintf("  %s %s: R-hat = %.4f\n", status, colnames(X)[j], rhat_mh_beta[j]))
}

cat("\n")

# ==============================================================================
# STEP 5: ALGORITHM COMPARISON
# ==============================================================================

cat("━"*80, "\n", sep="")
cat("STEP 5: ALGORITHM COMPARISON\n")
cat("━"*80, "\n\n", sep="")

compare_algorithms(
  gibbs_results = gibbs_results,
  mh_results = mh_results,
  param_names = colnames(X),
  output_dir = "../outputs/algorithm_comparison"
)

# ==============================================================================
# STEP 6: RESIDUAL DIAGNOSTICS
# ==============================================================================

cat("\n")
cat("━"*80, "\n", sep="")
cat("STEP 6: RESIDUAL DIAGNOSTICS\n")
cat("━"*80, "\n\n", sep="")

# Combine Gibbs chains
gibbs_beta_all <- do.call(rbind, gibbs_beta_chains)
gibbs_sigma2_all <- do.call(c, gibbs_sigma2_chains)

cat("Generating residual diagnostic plots...\n\n")

plot_residual_diagnostics(
  y_obs = y,
  X = X,
  beta_samples = gibbs_beta_all,
  sigma2_samples = gibbs_sigma2_all,
  output_path = "../outputs/gibbs_result/residual_diagnostics.png",
  model_name = "Gibbs Sampler"
)

# ==============================================================================
# STEP 7: POSTERIOR INFERENCE AND SUMMARY
# ==============================================================================

cat("\n")
cat("━"*80, "\n", sep="")
cat("STEP 7: POSTERIOR INFERENCE AND SUMMARY\n")
cat("━"*80, "\n\n", sep="")

# Posterior estimates
beta_mean <- colMeans(gibbs_beta_all)
sigma2_mean <- mean(gibbs_sigma2_all)

cat("Posterior Estimates:\n")
for (j in 1:p) {
  ci_lower <- quantile(gibbs_beta_all[, j], 0.025)
  ci_upper <- quantile(gibbs_beta_all[, j], 0.975)
  cat(sprintf("  %s: %.4f [%.4f, %.4f]\n", colnames(X)[j], beta_mean[j], ci_lower, ci_upper))
}
cat(sprintf("  σ²: %.4f\n\n", sigma2_mean))

# Model fit
y_pred <- X %*% beta_mean
r_squared <- 1 - sum((y - y_pred)^2) / sum((y - mean(y))^2)
rmse <- sqrt(mean((y - y_pred)^2))
mae <- mean(abs(y - y_pred))

cat("Model Fit Statistics:\n")
cat(sprintf("  R² = %.4f (%.1f%% variance explained)\n", r_squared, r_squared * 100))
cat(sprintf("  RMSE = $%.2f\n", rmse))
cat(sprintf("  MAE = $%.2f\n\n", mae))

# Save summary
output_dir <- "../outputs/final_summary"
dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)

# Posterior estimates table
posterior_summary <- data.frame(
  Parameter = c(colnames(X), "σ²"),
  Mean = c(beta_mean, sigma2_mean),
  SD = c(apply(gibbs_beta_all, 2, sd), sd(gibbs_sigma2_all)),
  CI_2.5 = c(apply(gibbs_beta_all, 2, function(x) quantile(x, 0.025)), quantile(gibbs_sigma2_all, 0.025)),
  CI_97.5 = c(apply(gibbs_beta_all, 2, function(x) quantile(x, 0.975)), quantile(gibbs_sigma2_all, 0.975))
)

write.csv(posterior_summary, file.path(output_dir, "posterior_summary.csv"), row.names = FALSE)

# Model fit table
model_fit <- data.frame(
  Statistic = c("R²", "RMSE", "MAE", "n", "p", "Effective_n"),
  Value = c(r_squared, rmse, mae, n, p, n)
)

write.csv(model_fit, file.path(output_dir, "model_fit.csv"), row.names = FALSE)

cat(sprintf("✓ Results saved to %s\n\n", output_dir))

# ==============================================================================
# FINAL SUMMARY
# ==============================================================================

analysis_end_time <- Sys.time()
total_time <- as.numeric(difftime(analysis_end_time, analysis_start_time, units = "secs"))

cat("\n")
cat("╔"*80, "\n", sep="")
cat("║ ANALYSIS COMPLETE\n")
cat("╚"*80, "\n", sep="")
cat(sprintf("Total analysis time: %.2f seconds (%.1f minutes)\n\n", total_time, total_time/60))

cat("Summary:\n")
cat(sprintf("  ✓ Data: %d observations, %d predictors\n", n, p-1))
cat(sprintf("  ✓ Algorithms: Gibbs + Metropolis-Hastings\n"))
cat(sprintf("  ✓ Chains: %d per algorithm (%d iterations each)\n", n_chains, n_iter))
cat(sprintf("  ✓ Convergence: %s\n", 
            ifelse(all(c(rhat_gibbs_beta, rhat_mh_beta) < 1.1), 
                   "All chains converged", 
                   "Check diagnostics")))
cat(sprintf("  ✓ Model fit: R² = %.3f\n\n", r_squared))

cat("Output directories:\n")
cat("  - ../outputs/gibbs_result/\n")
cat("  - ../outputs/mh_baseline/\n")
cat("  - ../outputs/algorithm_comparison/\n")
cat("  - ../outputs/final_summary/\n\n")

cat("="*80, "\n", sep="")
cat("All results saved. Analysis pipeline complete!\n")
cat("="*80, "\n", sep="")

File: python/scripts/run_full_analysis.py
#!/usr/bin/env python
"""
Complete analysis runner.
"""

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from pathlib import Path
from sklearn.preprocessing import StandardScaler
import sys

from data_preprocessing import preprocess_data
from model_setup import create_design_matrix, run_ols_baseline
from gibbs_sampling import gibbs_lm, beta_trace_plot, sigma2_trace_plot
from convergence_detection import acf_plot_beta, acf_plot_sigma2
from posterior_inference import posterior_predictive, ppc_plot, PPC_density_overlay, ppc_residual_plot
from mcmc_diagnostics import rhat, ess_geyer

sns.set_style("whitegrid")
plt.rcParams['figure.figsize'] = (12, 8)

print("="*80)
print("Bayesian MCMC Insurance Regression - Full Analysis")
print("="*80)

print("\n[1/7] Loading data...")

df = pd.read_csv('../../data/expenses_cleaned.csv')
print(f"   Data shape: {df.shape}")
print(f"   Columns: {list(df.columns)}")

y = df['charges'].values
X = df.drop('charges', axis=1).values
n, p = X.shape

X_with_intercept = np.column_stack([np.ones(n), X])
feature_names = ['Intercept'] + list(df.drop('charges', axis=1).columns)

print(f"   n = {n} observations")
print(f"   p = {p} predictors (+ intercept)")
print(f"   Features: {feature_names}")

print("\n[2/7] Running Gibbs sampling...")
print("   Parameters: 50,000 iterations, 10,000 warmup, 3 chains")

# Run Gibbs sampler
chains_baseline = gibbs_lm(
    y=y,
    X=X_with_intercept,
    n_iter=50000,
    warmup=10000,
    n_chains=3,
    seed=42
)

# Extract samples
beta_chains = [chain['beta'] for chain in chains_baseline]
sigma2_chains = [chain['sigma2'] for chain in chains_baseline]

print("   ✓ Sampling complete")

# ============================================================================
# 3. CONVERGENCE DIAGNOSTICS
# ============================================================================
print("\n[3/7] Computing convergence diagnostics...")

# Create output directory
output_dir = Path('../outputs/baseline_model')
output_dir.mkdir(parents=True, exist_ok=True)

# Compute R-hat using manual implementation
rhat_beta = []
for j in range(X_with_intercept.shape[1]):
    chains_j = np.array([chain[:, j] for chain in beta_chains])  # shape: (n_chains, n_samples)
    rhat_beta.append(rhat(chains_j))

rhat_sigma2 = rhat(np.array(sigma2_chains))

print(f"   R-hat (beta): {[f'{r:.4f}' for r in rhat_beta]}")
print(f"   R-hat (sigma2): {rhat_sigma2:.4f}")

# Check convergence
converged = all(r < 1.1 for r in rhat_beta) and rhat_sigma2 < 1.1
print(f"   Convergence: {'✓ CONVERGED' if converged else '✗ NOT CONVERGED'}")

# Compute ESS
ess_dir = output_dir / 'ESS_tables'
ess_dir.mkdir(exist_ok=True)

ess_beta = []
for j in range(X_with_intercept.shape[1]):
    chains_j = np.array([chain[:, j] for chain in beta_chains])  # shape: (n_chains, n_samples)
    ess_beta.append(ess_geyer(chains_j))

# Save ESS table
ess_df = pd.DataFrame({
    'Parameter': feature_names,
    'ESS': ess_beta
})
ess_df.to_csv(ess_dir / 'ESS_beta.txt', index=False, sep='\t')
print(f"   ESS (beta): {[f'{e:.0f}' for e in ess_beta]}")

# ============================================================================
# 4. POSTERIOR SUMMARIES
# ============================================================================
print("\n[4/7] Computing posterior summaries...")

# Combine chains
beta_combined = np.vstack(beta_chains)
sigma2_combined = np.concatenate(sigma2_chains)

# Compute summaries manually
mean_beta = np.mean(beta_combined, axis=0)
median_beta = np.median(beta_combined, axis=0)
sd_beta = np.std(beta_combined, axis=0, ddof=1)
ci_lower = np.percentile(beta_combined, 2.5, axis=0)
ci_upper = np.percentile(beta_combined, 97.5, axis=0)

print("\n   Posterior Estimates:")
print("   " + "-"*70)
print(f"   {'Parameter':<20} {'Mean':>10} {'Median':>10} {'SD':>10}")
print("   " + "-"*70)
for i, name in enumerate(feature_names):
    print(f"   {name:<20} {mean_beta[i]:>10.4f} "
          f"{median_beta[i]:>10.4f} {sd_beta[i]:>10.4f}")
print("   " + "-"*70)

# Save posterior estimates
posterior_df = pd.DataFrame({
    'Parameter': feature_names,
    'Mean': mean_beta,
    'Median': median_beta,
    'SD': sd_beta,
    'CI_lower': ci_lower,
    'CI_upper': ci_upper
})
posterior_df.to_csv(output_dir / 'posterior_estimates.csv', index=False)

# ============================================================================
# 5. GENERATE TRACE PLOTS
# ============================================================================
print("\n[5/7] Generating trace plots...")

# Note: beta_trace_plot expects list of arrays, not feature_names parameter
beta_trace_plot(beta_chains, model_name='baseline_model', plot_dir='../outputs')
sigma2_trace_plot(sigma2_chains, model_name='baseline_model', plot_dir='../outputs')

print("   ✓ Trace plots saved")

# ============================================================================
# 6. RESIDUAL DIAGNOSTICS
# ============================================================================
print("\n[6/7] Computing residual diagnostics...")

# Use posterior mean for predictions
y_pred = X_with_intercept @ mean_beta
residuals = y - y_pred

# Simple residual plot
fig, axes = plt.subplots(1, 2, figsize=(14, 5))

# Residuals vs fitted
axes[0].scatter(y_pred, residuals, alpha=0.5)
axes[0].axhline(y=0, color='r', linestyle='--')
axes[0].set_xlabel('Fitted values')
axes[0].set_ylabel('Residuals')
axes[0].set_title('Residuals vs Fitted')

# Q-Q plot
from scipy import stats as sp_stats
sp_stats.probplot(residuals, dist="norm", plot=axes[1])
axes[1].set_title('Normal Q-Q Plot')

plt.tight_layout()
plt.savefig(output_dir / 'residual_diagnostics.png', dpi=300, bbox_inches='tight')
plt.close()

print("   ✓ Residual diagnostics saved")

# ============================================================================
# 7. POSTERIOR PREDICTIVE CHECKS
# ============================================================================
print("\n[7/7] Posterior predictive checks...")

# Sample from posterior predictive
n_samples = 1000
n_obs = len(y)
y_rep = np.zeros((n_samples, n_obs))

# Random sample from posterior
idx = np.random.choice(len(beta_combined), size=n_samples, replace=False)

for i, j in enumerate(idx):
    beta_sample = beta_combined[j]
    sigma2_sample = sigma2_combined[j]
    y_rep[i] = X_with_intercept @ beta_sample + \
               np.random.normal(0, np.sqrt(sigma2_sample), n_obs)

# Compute test statistics
T_obs = np.mean(y)
T_rep = np.mean(y_rep, axis=1)
p_value = np.mean(T_rep >= T_obs)

print(f"   Posterior predictive p-value (mean): {p_value:.4f}")

# Plot PPC
ppc_dir = output_dir / 'PPC'
ppc_dir.mkdir(exist_ok=True)

fig, ax = plt.subplots(figsize=(10, 6))
ax.hist(T_rep, bins=50, alpha=0.7, edgecolor='black', label='Replicated data')
ax.axvline(T_obs, color='red', linestyle='--', linewidth=2, label='Observed data')
ax.set_xlabel('Mean charges', fontsize=12)
ax.set_ylabel('Frequency', fontsize=12)
ax.set_title('Posterior Predictive Check: Mean', fontsize=14)
ax.legend()
plt.tight_layout()
plt.savefig(ppc_dir / 'ppc_mean.png', dpi=300, bbox_inches='tight')
plt.close()

print("   ✓ PPC plots saved")

# ============================================================================
# SUMMARY
# ============================================================================
print("\n" + "="*80)
print("ANALYSIS COMPLETE")
print("="*80)
print(f"\nResults saved to: {output_dir.absolute()}")
print("\nKey findings:")
print(f"  - All chains converged (R-hat < 1.1): {converged}")
print(f"  - Effective sample sizes: {int(np.mean(ess_beta)):.0f} (average)")
print(f"  - Strongest predictor: {feature_names[np.argmax(np.abs(mean_beta))]} "
      f"(β = {mean_beta[np.argmax(np.abs(mean_beta))]:.3f})")
print(f"  - Model R²: {1 - np.var(residuals)/np.var(y):.4f}")
print("\n" + "="*80)
